In [6]:
import pandas as pd
import numpy as np

# 1. Load the dataset
df = pd.read_csv('/content/Telco_Cusomer_Churn.csv')

# 2. Structural Data Type Cleaning
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'].str.strip(), errors='coerce')
df['TotalCharges'] = df['TotalCharges'].fillna(0.0)

# --- ADVANCED BUSINESS KPI EXTRACTION ---

# Total Revenue Leakage Metrics
total_active_subscribers = df[df['Churn'] == 'No'].shape[0]
total_churned_subscribers = df[df['Churn'] == 'Yes'].shape[0]
global_churn_rate = (total_churned_subscribers / len(df)) * 100

monthly_revenue_at_risk = df[df['Churn'] == 'Yes']['MonthlyCharges'].sum()
historical_lost_revenue = df[df['Churn'] == 'Yes']['TotalCharges'].sum()

# Print out a clean, high-impact Executive Health Check Table
print("="*70)
print("             🚨 EXECUTIVE REVENUE & RETENTION AUDIT REPORT 🚨             ")
print("="*70)
print(f"📊 CUSTOMER BASEMETRICS:")
print(f"  • Total Customers Tracked       : {len(df):,}")
print(f"  • Active Retained Subscribers   : {total_active_subscribers:,} ({100 - global_churn_rate:.1f}%)")
print(f"  • Terminated (Churned) Accounts : {total_churned_subscribers:,} ({global_churn_rate:.1f}%)")
print("-"*70)
print(f"💰 FINANCIAL IMPACT MATRIX:")
print(f"  • Total Historical Revenue Lost : ${historical_lost_revenue:,.2f}")
print(f"  • Monthly Recurring Revenue (MRR) Lost Today: ${monthly_revenue_at_risk:,.2f}")
print("="*70)

# Display structural dataset summaries below the financials
print("\n📋 DATA INFRASTRUCTURE HEALTH:")
print(f"• Shape: {df.shape[0]} rows | {df.shape[1]} columns")
print(f"• Missing data points resolved: {df['TotalCharges'].isnull().sum()}")
print(f"• Duplicated entities dropped  : {df.duplicated().sum()}")
print("-"*70)
df.head(3)

             🚨 EXECUTIVE REVENUE & RETENTION AUDIT REPORT 🚨             
📊 CUSTOMER BASEMETRICS:
  • Total Customers Tracked       : 7,043
  • Active Retained Subscribers   : 5,174 (73.5%)
  • Terminated (Churned) Accounts : 1,869 (26.5%)
----------------------------------------------------------------------
💰 FINANCIAL IMPACT MATRIX:
  • Total Historical Revenue Lost : $2,862,926.90
  • Monthly Recurring Revenue (MRR) Lost Today: $139,130.85

📋 DATA INFRASTRUCTURE HEALTH:
• Shape: 7043 rows | 21 columns
• Missing data points resolved: 0
• Duplicated entities dropped  : 0
----------------------------------------------------------------------


,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.50,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes


## 📁 Stage 1: Advanced Structural Quality Control & Financial Baseline Audit

### 🛠️ Data Integrity & Transformation Methodology
To prevent metric distortion during downstream retention visualization, we initiated an explicit cleaning protocol:
* **Datatype Coercion:** The feature `TotalCharges` was natively cataloged as a generic string object due to whitespace omissions. These profiles were programmatically forced into an analytical `float64` framework.
* **Preservation Imputation:** 11 active accounts showcased null value bounds in their lifetime pricing tallies. Investigation verified these accounts held a `tenure` of 0 months. Rather than dropping valid customer matrices, these boundaries were imputed to `$0.00` to reflect zero billing history.

### 💰 Executive Financial Breakdown (The Baseline Story)
Analyzing the fundamental split of the data highlights an immediate, high-priority risk framework for business leadership:

1. **Severe Volatility Index:** The database indicates a baseline **26.5% Global Churn Rate**. Losing over a quarter of the total customer base points to an active loyalty leakage that threatens scalable unit economics.
2. **Quantifiable Financial Bleeding:**
   * **Historical Sunk Capital:** Terminated relationships have directly resulted in **$2,862,926.90** in historically lost gross revenue.
   * **Active Operational Drainage:** The current batch of churned accounts has stripped **$139,130.85 in Monthly Recurring Revenue (MRR)** straight out of the active baseline ecosystem.
3. **Analytical Target Focus:** Moving into the segmentation stage, our core goal will look past simple demographics to identify precisely which product models or subscription terms are triggering this $139K monthly revenue drain.